File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ValidateArray.java
package io.github.brenbar.uapi.internal;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.github.brenbar.uapi.internal.types.UTypeDeclaration;

import static io.github.brenbar.uapi.internal.GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure;
import static io.github.brenbar.uapi.internal.Prepend.prepend;
import static io.github.brenbar.uapi.internal.types.UArray._ARRAY_NAME;

public class ValidateArray {
    public static List<ValidationFailure> validateArray(Object value, Map<String, Object> select, String fn,
            List<UTypeDeclaration> typeParameters,
            List<UTypeDeclaration> generics) {
        if (value instanceof final List l) {
            final var nestedTypeDeclaration = typeParameters.get(0);

            final var validationFailures = new ArrayList<ValidationFailure>();
            for (var i = 0; i < l.size(); i += 1) {
                final var element = l.get(i);
                final var nestedValidationFailures = nestedTypeDeclaration.validate(element, select, fn, generics);
                final var index = i;

                final var nestedValidationFailuresWithPath = new ArrayList<ValidationFailure>();
                for (var f : nestedValidationFailures) {
                    final List<Object> finalPath = prepend(index, f.path);
                    nestedValidationFailuresWithPath.add(new ValidationFailure(finalPath, f.reason,
                            f.data));
                }

                validationFailures.addAll(nestedValidationFailuresWithPath);
            }

            return validationFailures;
        } else {
            return getTypeUnexpectedValidationFailure(List.of(), value,
                    _ARRAY_NAME);
        }
    }
}

File: lib/py/telepact/internal/validation/ValidateArray.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING
from ..types.TArray import _ARRAY_NAME
from ...internal.validation.ValidationFailure import ValidationFailure
from ...internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure

if TYPE_CHECKING:
    from ...internal.validation.ValidateContext import ValidateContext
    from ..types.TTypeDeclaration import TTypeDeclaration


def validate_array(value: object,
                   type_parameters: list['TTypeDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
    if type(value) is not list and not isinstance(value, list):
        return get_type_unexpected_validation_failure([], value, _ARRAY_NAME)

    nested_type_declaration = type_parameters[0]

    validation_failures = []
    for i, element in enumerate(value):
        ctx.path.append("*")

        nested_validation_failures = nested_type_declaration.validate(
            element, ctx)
        index = i

        ctx.path.pop()

        if not nested_validation_failures:
            continue

        for f in nested_validation_failures:
            validation_failures.append(
                ValidationFailure([index] + f.path, f.reason, f.data))

    return validation_failures
